### PR TITLE
environment.shellAliases: change default behavior

### PR DESCRIPTION
**Pull request by mnacamura**
- Do not override environment.shellAliases in `nixos/modules/programs/shell.nix`.
- `programs.{bash,zsh,fish}.shellAliases` override `environment.shellAliases` for each shell.

###### Motivation for this change

Hopefully improve the situation in #36282.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/config/shells-environment.nix [hunk 1]
```diff
@@ -107,7 +107,11 @@ in
     };
 
     environment.shellAliases = mkOption {
-      default = {};
+      default = {
+        ls = "ls --color=tty";
+        ll = "ls -l";
+        l  = "ls -alh";
+      };
```
Comment (teto) [inline nixos/modules/config/shells-environment.nix]: why change the defaults ?

File: nixos/modules/config/shells-environment.nix [hunk 2]
```diff
@@ -107,7 +107,11 @@ in
     };
 
     environment.shellAliases = mkOption {
-      default = {};
```
Comment (FRidh) [inline nixos/modules/config/shells-environment.nix]: Are these defaults intentional?

Comment (mnacamura) [reply nixos/modules/config/shells-environment.nix]: I just moved these aliases that were in `nixos/modules/programs/shell.nix` to here without any specific reason. Empty default is fine or better I think though.

**Comment by mnacamura** [comment]
Made default `environment.shellAliases` empty.

File: nixos/modules/config/shells-environment.nix (hunk 2 above)
Comment (mnacamura) [reply nixos/modules/config/shells-environment.nix]: Never mind. It just reflected my preference.

**Comment by mnacamura** [comment]
Probably it could be better, for example, if
```
environment.shellAliases = {
  ls = "ls -a";
}

programs.bash.shellAliases = {
  ls = "ls -h";
}
```
then `ls` is expanded to `ls -ah` in bash?

File: nixos/modules/programs/shell.nix [hunk 3]
```diff
@@ -8,12 +8,6 @@ with lib;
 
   config = {
 
-    environment.shellAliases =
-      { ls = "ls --color=tty";
-        ll = "ls -l";
-        l  = "ls -alh";
-      };
-
```
Comment (Infinisil) [inline nixos/modules/programs/shell.nix]: Not entirely sure about removing those, they have been there for a long time. I don't personally use them, but maybe some people do. Do other distros have any default aliases?

Also, the cost of keeping these around with the changes in this PR isn't very high, so we might as well just leave them in for the sake of not annoying anybody.

File: nixos/modules/programs/bash/bash.nix [hunk 4]
```diff
@@ -33,7 +33,7 @@ let
   '';
 
   bashAliases = concatStringsSep "\n" (
-    mapAttrsFlatten (k: v: "alias ${k}=${escapeShellArg v}") cfg.shellAliases
+    mapAttrsFlatten (k: v: "alias ${k}=${escapeShellArg v}") (cfge.shellAliases // cfg.shellAliases)
```
Comment (Infinisil) [inline nixos/modules/programs/bash/bash.nix]: I think a better way to implement this is to have
```nix
{
  config.programs.bash.shellAliases = mapAttrs (name: mkDefault) cfge.shellAliases;
}
```

This should then have the `//` override behavior automatically through the module system. Also this allows the user to query all bash aliases by evaluating the option.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/shell.nix (hunk 3 above)
Comment (mnacamura) [reply nixos/modules/programs/shell.nix]: Okey. Moved them to the default of `environment.shellAliases`.

File: nixos/modules/config/shells-environment.nix (hunk 2 above)
Comment (mnacamura) [reply nixos/modules/config/shells-environment.nix]: Reverted the default aliases. See https://github.com/NixOS/nixpkgs/pull/44441#discussion_r224580308

File: nixos/modules/programs/bash/bash.nix (hunk 4 above)
Comment (mnacamura) [reply nixos/modules/programs/bash/bash.nix]: Hmm... By your change, although defined in `/etc/nixos/configuration.nix`, bash aliases that have the same names as defined in `environment.shellAliases` seem not appearing in `programs.bash.shellAliases`.

**Comment by Infinisil** [comment]
Ah yeah, there's a problem with the original type. This works:
```diff
diff --git a/nixos/modules/config/shells-environment.nix b/nixos/modules/config/shells-environment.nix
index 5ecc3cfd207..a279120e125 100644
--- a/nixos/modules/config/shells-environment.nix
+++ b/nixos/modules/config/shells-environment.nix
@@ -108,18 +108,13 @@ in
     };
 
     environment.shellAliases = mkOption {
-      default = {
-        ls = "ls --color=tty";
-        ll = "ls -l";
-        l  = "ls -alh";
-      };
       example = { ll = "ls -l"; };
       description = ''
         An attribute set that maps aliases (the top level attribute names in
         this option) to command strings or directly to build outputs. The
         aliases are added to all users' shells.
       '';
-      type = types.attrs; # types.attrsOf types.stringOrPath;
+      type = with types; attrsOf (either str path);
     };
 
     environment.binsh = mkOption {
@@ -161,6 +156,12 @@ in
     # terminal instead of logging out of X11).
     environment.variables = config.environment.sessionVariables;
 
+    environment.shellAliases = mapAttrs (name: mkDefault) {
+      ls = "ls --color=tty";
+      ll = "ls -l";
+      l  = "ls -alh";
+    };
+
     environment.etc."shells".text =
       ''
         ${concatStringsSep "\n" (map utils.toShellPath cfg.shells)}
diff --git a/nixos/modules/programs/bash/bash.nix b/nixos/modules/programs/bash/bash.nix
index 62c63f2aeb3..aa524a333ee 100644
--- a/nixos/modules/programs/bash/bash.nix
+++ b/nixos/modules/programs/bash/bash.nix
@@ -33,7 +33,7 @@ let
   '';
 
   bashAliases = concatStringsSep "\n" (
-    mapAttrsFlatten (k: v: "alias ${k}=${escapeShellArg v}") (cfge.shellAliases // cfg.shellAliases)
+    mapAttrsFlatten (k: v: "alias ${k}=${escapeShellArg v}") cfg.shellAliases
   );
 
 in
@@ -64,7 +64,7 @@ in
           Set of aliases for bash shell, which overrides <option>environment.shellAliases</option>.
           See <option>environment.shellAliases</option> for an option format description.
         '';
-        type = types.attrs; # types.attrsOf types.stringOrPath;
+        type = with types; attrsOf (either str path);
       };
 
       shellInit = mkOption {
@@ -125,6 +125,8 @@ in
 
     programs.bash = {
 
+      shellAliases = mapAttrs (name: mkDefault) cfge.shellAliases;
+
       shellInit = ''
         if [ -z "$__NIXOS_SET_ENVIRONMENT_DONE" ]; then
             . ${config.system.build.setEnvironment}
```

(the same for fish and zsh)

**Comment by mnacamura** [comment]
The above patch works. Thank you!

Another problem that concerns me is how to remove the default aliases defined in `environment.shellAliases`. I mean, for example, how to remove `l="ls -alh"` if it is not necessary.

**Comment by Infinisil** [comment]
Yeah, usually that's done by allowing `null` values (prepending the type with `nullOr`) and handling that as non-presence when generating. Feel free to add support for that as well.

**Comment by mnacamura** [comment]
All set.

File: nixos/modules/config/shells-environment.nix [hunk 5]
```diff
@@ -108,14 +108,18 @@ in
...
+      default = mapAttrs (name: mkDefault) {
+        ls = "ls --color=tty";
+        ll = "ls -l";
```
Comment (Infinisil) [inline nixos/modules/config/shells-environment.nix]: I don't think this should work like that. It needs to be in the config section. How it's currently will disable all default aliases if a user adds any of their own.

Comment (mnacamura) [reply nixos/modules/config/shells-environment.nix]: Oh, right. Sorry.

**Review by infinisil — APPROVED** [review]

**Comment by Infinisil** [comment]
This now works as follows:
- `programs.{bash,zsh,fish}.shellAliases` take defaults from `environment.shellAliases`
- Defaults for `environment.shellAliases` are `l`, `ll` and `ls`, you can override these by setting `enviromnent.shellAliases.{l,ll,ls}` for all shells, or `programs.{bash,zsh,fish}.shellAliases.{l,ll,ls}` for just a specific one.

Solves all issues mentioned in https://github.com/NixOS/nixpkgs/issues/36282

**Pull request merged**